server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -2464,203 +2465,255 @@ public boolean assertAllListenersResolved() {
-                // completed snapshots do not require any updates so we just add them to the new list and keep going
+
+        // builder for updated shard snapshot status mappings for the #current snapshot entry if any could be computed
+        ImmutableOpenMap.Builder<ShardId, ShardSnapshotStatus> shardsBuilder = null;
+
+        // builder for updated shard clone status mappings for the #current snapshot entry if any could be computed
+        ImmutableOpenMap.Builder<RepositoryShardId, ShardSnapshotStatus> clonesBuilder = null;
+
+        // index lookup (index name -> IndexId) cache for the #current snapshot entry used to translate from ShardId to RepositoryShardId
+        private Map<String, IndexId> indicesLookup = null;
+
+        // iterator containing the tasks yet to be applied to #current
+        private Iterator<ShardSnapshotUpdate> currentTaskIterator;
+
+        // current snapshot entry to update
+        private SnapshotsInProgress.Entry current;
+
+        SnapshotShardsUpdateContext(ClusterState currentState, List<ShardSnapshotUpdate> tasks) {
+            this.currentState = currentState;
+            unconsumedTasks = new ArrayList<>(tasks);
+        }
+
+        public ClusterState updatedState() {
+            if (changedCount > 0) {
+                logger.trace("changed cluster state triggered by [{}] snapshot state updates and resulted in starting " +
+                        "[{}] shard snapshots", changedCount, startedCount);
+                return ClusterState.builder(currentState).putCustom(SnapshotsInProgress.TYPE, SnapshotsInProgress.of(entries)).build();
+            }
+            return currentState;
+        }
+
+        void applyToEntry(SnapshotsInProgress.Entry entry) {
+            // Completed snapshots do not require any updates so we just add them to the output list and keep going.
+            // Also we short circuit if there are no more unconsumed tasks to apply.
+            if (entry.state().completed() || unconsumedTasks.isEmpty()) {
                 entries.add(entry);
-                continue;
+                return;
             }
-            ImmutableOpenMap.Builder<ShardId, ShardSnapshotStatus> shards = null;
-            ImmutableOpenMap.Builder<RepositoryShardId, ShardSnapshotStatus> clones = null;
-            Map<String, IndexId> indicesLookup = null;
-            // inner loop over all the shard updates that are potentially applicable to the current snapshot entry
-            for (Iterator<ShardSnapshotUpdate> iterator = unconsumedTasks.iterator(); iterator.hasNext(); ) {
-                final ShardSnapshotUpdate updateSnapshotState = iterator.next();
-                final Snapshot updatedSnapshot = updateSnapshotState.snapshot;
-                final String updatedRepository = updatedSnapshot.getRepository();
-                if (entry.repository().equals(updatedRepository) == false) {
+
+            // reset internal state for new entry
```

Comment:
Hmm this is a bit suspicious, it indicates we are confusing per-entry-computation state with per-application state in this class. I would prefer us to have another class representing the per-entry context here, with these fields belonging to that inner context.

Reply:
++ introduced a class to hold all the per entry fields now

Reply:
How about making it non-static and moving all the methods onto that context too? Seems better than reaching inside it to grab the `entry` and `iterator` private fields etc.

Reply:
Right did that :) that does look a lot cuter now.